[SYCL][Driver]Enable multiple third-party targets for SYCL AOT.

srividya-sundaram opened the pull request:
This patch enables `-fsycl-targets` option to support a comma-separated list of `NVidia` and `AMD` Processor Graphics values as target triple values for the SYCL AOT flow.

**Example:**
```
clang++  -###  -fsycl  -fsycl-targets=nvidia_gpu_sm_60,nvidia_gpu_sm_70  syclfile.cpp
clang++  -###  -fsycl  -fsycl-targets=amd_gpu_gfx90a,amd_gpu_gfx90c      syclfile.cpp

```

mdtoguchi commented on clang/test/Driver/sycl-offload-old-model.c:
```diff
@@ -861,3 +861,15 @@
+// RUN:   | FileCheck -check-prefixes=CHK-MACRO-GFX90A,CHK-MACRO-GFX90C %s
+// CHK-MACRO-GFX90A: clang{{.*}} "-fsycl-is-device"{{.*}} "-D__SYCL_TARGET_AMD_GPU_GFX90A__"{{.*}}
+// CHK-MACRO-GFX90C: clang{{.*}} "-fsycl-is-device"{{.*}} "-D__SYCL_TARGET_AMD_GPU_GFX90C__"{{.*}}
+
```
We should have something that verifies this also is working correctly with the new offloading model.

mdtoguchi reviewed (approved):
LGTM - are there plans on updating documentation to allow for multiple target support?

mdtoguchi commented on sycl/doc/UsersManual.md:
```diff
@@ -45,8 +45,8 @@ and not recommended to use in production environment.
-    support are accepted, providing a streamlined interface for AOT. Only one of
-    these values at a time is supported.
+    support are accepted, providing a streamlined interface for AOT.
+    A comma-separated list of valid Intel,NVidia and AMD Processor Graphics values are supported.
```
```suggestion
    A comma-separated list of valid Intel, NVIDIA and AMD Processor Graphics
    values are supported.
```
I don't know what the preferred way of representing `NVIDIA` is, but we should be consistent with usage (here and line 47)

maarquitos14 reviewed (approved):
LGTM, just a nit.

dm-vodopyanov commented on sycl/test/check_device_code/hip/atomic/amdgpu_unsafe_atomics.cpp:
```diff
@@ -1,7 +1,7 @@
-// RUN: %clangxx -fsycl -fsycl-targets=amd_gpu_gfx906 %s -S -emit-llvm -o - | FileCheck %s --check-prefixes=CHECK,CHECK-SAFE
-// RUN: %clangxx -fsycl -fsycl-targets=amd_gpu_gfx906 %s -mllvm --amdgpu-oclc-unsafe-int-atomics=true -S -emit-llvm -o - | FileCheck %s --check-prefixes=CHECK,CHECK-UNSAFE
-// RUN: %clangxx -fsycl -fsycl-targets=amd_gpu_gfx90a %s -mllvm --amdgpu-oclc-unsafe-fp-atomics=true  -mllvm --amdgpu-oclc-unsafe-int-atomics=true -S -emit-llvm -o - | FileCheck %s --check-prefixes=CHECK,CHECK-UNSAFE-FP
+// RUN: %clangxx -fsycl -fsycl-targets=amdgcn-amd-amdhsa %s -S -emit-llvm -o - | FileCheck %s --check-prefixes=CHECK,CHECK-SAFE
+// RUN: %clangxx -fsycl -fsycl-targets=amdgcn-amd-amdhsa %s -mllvm --amdgpu-oclc-unsafe-int-atomics=true -S -emit-llvm -o - | FileCheck %s --check-prefixes=CHECK,CHECK-UNSAFE
+// RUN: %clangxx -fsycl -fsycl-targets=amdgcn-amd-amdhsa %s -mllvm --amdgpu-oclc-unsafe-fp-atomics=true  -mllvm --amdgpu-oclc-unsafe-int-atomics=true -S -emit-llvm -o - | FileCheck %s --check-prefixes=CHECK,CHECK-UNSAFE-FP
```
I think the idea is that these values must be aligned with values from sycl_ext_oneapi_device_architecture ([link](https://github.com/intel/llvm/blob/7d063d229ea44608abae49044ef718ddc26348b3/sycl/include/sycl/ext/oneapi/experimental/device_architecture.def#L118-L121)).
Change in this PR can confuse our customers as there is no "amdgcn-amd-amdhsa" in sycl_ext_oneapi_device_architecture

srividya-sundaram replied:
@dm-vodopyanov 

The test files were updated because these [lines](https://github.com/intel/llvm/blob/sycl/sycl/test/lit.cfg.py#L173) from [lit.cfg.py](https://github.com/intel/llvm/blob/sycl/sycl/test/lit.cfg.py) were adding `"-Xsycl-target-backend=amdgcn-amd-amdhsa", "--offload-arch=gfx906"` options in addition to the existing driver options in the RUN command in the test files.

This was resulting in either :
1.  two similar targets, causing duplication error `(error: Duplicate targets are not allowed)` or 
2.  two different targets causing : `clang: error: cannot specify -o when generating multiple output files`

Additionally, all the matrix related tests have `joint_matrix` usage, where the `joint_matrix` API is only supported by the Intel, CUDA and HIP (`GFX90A`) backends.

Hence I updated the default AMD GPU device to be `GFX90A`.

dm-vodopyanov replied:
Ok, thanks for explanation, LGTM

AlexeySachkov replied:
Why don't we just drop that `lit.cfg.py` hardcode for `--offload-arch` in favor of `--fsycl-targets`?

srividya-sundaram replied:
When I removed the hardcoded check, there were about ~60 tests failing with the error : `clang: error: missing AMDGPU architecture for SYCL offloading; specify it with '-Xsycl-target-backend=amdgcn-amd-amdhsa --offload-arch=<arch-name>'`.
Hence they were not removed.

dm-vodopyanov commented:
From the PR description: 
> ```
> clang++  -###  -fsycl  -fsycl-targets=nvidia_gpu_sm_60, nvidia_gpu_sm_70  syclfile.cpp
> clang++  -###  -fsycl  -fsycl-targets=amd_gpu_gfx90a, amd_gpu_gfx90c      syclfile.cpp
> ```

I assume there should be no space between values in `-fsycl-targets` option? If not, please update the description

srividya-sundaram commented:
> From the PR description:
> 
> > ```
> > clang++  -###  -fsycl  -fsycl-targets=nvidia_gpu_sm_60, nvidia_gpu_sm_70  syclfile.cpp
> > clang++  -###  -fsycl  -fsycl-targets=amd_gpu_gfx90a, amd_gpu_gfx90c      syclfile.cpp
> > ```
> 
> I assume there should be no space between values in `-fsycl-targets` option? If not, please update the description

I added some extra space for readability purposes but I have removed the spaces now in the description.

AlexeySachkov commented on sycl/doc/UsersManual.md:
```diff
@@ -45,8 +45,8 @@ and not recommended to use in production environment.
-    support are accepted, providing a streamlined interface for AOT. Only one of
-    these values at a time is supported.
+    support are accepted, providing a streamlined interface for AOT.
+    A comma-separated list of valid Intel, NVIDIA and AMD Processor Graphics values are supported.
```
What is the behavior if user combines "special" target triples with "generic" target triples defined above?

I.e. something like `-fsycl-targets=spir64_gen,intel_gpu_ptl_u`?

srividya-sundaram replied:
Ideally, for,  `-fsycl-targets=spir64_gen,intel_gpu_ptl_u`, we would provide the target GPU device via 
`-Xsycl-target-backend=spir64_gen "-device <device-name>"` for `spir64_gen` AOT targets.

For the example, `-fsycl-targets=spir64_gen,intel_gpu_ptl_u`, ptl_u is passed as the target device for SYCL device compilation and ocloc call. Since no specific GPU device is provided for spir64_gen AOT target, no device info is passed to the `SYCL` device compilation step and the `ocloc` call.

Example:
**device = ptl_u**
```
"clang-20" "-cc1" "-triple" "spir64_gen-unknown-unknown" "-fsycl-is-device" "-D__SYCL_TARGET_INTEL_GPU_PTL_U__" ...
"ocloc"  ...."-output_no_suffix" "-spirv_input" "-device" "ptl_u"
```

**No device info passed (fsycl-targets=spir64_gen)**

```
"clang-20" "-cc1" "-triple" "spir64_gen-unknown-unknown" "-fsycl-is-device" ....
"ocloc" .... "-output_no_suffix" "-spirv_input"
```

AlexeySachkov replied:
> Ideally, for, `-fsycl-targets=spir64_gen,intel_gpu_ptl_u`, we would provide the target GPU device via
`-Xsycl-target-backend=spir64_gen "-device <device-name>"` for `spir64_gen` AOT targets.

Ok, so I understand it as this combination works (assuming that `-device` was passed for `spir64_gen` target). I was asking because I think that it is important to support it as a mechanism to do gradual transition from `spir64_gen` to special targets for our users.

srividya-sundaram commented:
@intel/llvm-gatekeepers Hello, this is ready for merge.Thanks!
